### PR TITLE
Fall back to opaque types rather than panicking on parse failure

### PR DESCRIPTION
**Pull request by fitzgen**
Getting closer to figuring out some of the other template related issues in clang 4.0, but not quite ready to land them yet. Figure this should probably land in the meantime. This is just a better fallback in the face of the unknown for panics that we've had reports of in the wild, but which I haven't had time to creduce.

r? @emilio

### CODE REVIEW

**Comment by fitzgen** [comment]
Potentially solves https://github.com/servo/rust-bindgen/issues/584 but I haven't tried to reduce those failures yet.

File: src/ir/context.rs [hunk 1]
```diff
@@ -424,7 +424,9 @@ impl<'ctx> BindgenContext<'ctx> {
         for (id, ty, loc, parent_id) in typerefs {
             let _resolved = {
                 let resolved = Item::from_ty(&ty, loc, parent_id, self)
-                    .expect("What happened?");
+                    .unwrap_or_else(|_| {
+                        Item::new_opaque_type(self.next_item_id(), &ty, self)
```
Comment (emilio) [inline src/ir/context.rs]: I believe we want to leave a warning here or something.

File: src/ir/ty.rs [hunk 2]
```diff
@@ -1093,12 +1093,14 @@ impl Type {
...
-                                let complex = CompInfo::from_ty(potential_id,
-                                                                ty,
-                                                                Some(location),
-                                                                ctx)
-                                    .expect("C'mon");
-                                TypeKind::Comp(complex)
+                                if let Ok(complex) = CompInfo::from_ty(potential_id,
+                                                                       ty,
+                                                                       Some(location),
+                                                                       ctx) {
```
Comment (emilio) [inline src/ir/ty.rs]: Perhaps using `match` + a separate variable may look a bit cleaner?

```rust
let complex = CompInfo::from_ty(..);
match complex {
    Ok(complex) => TypeKind::Comp(complex),
    Err(..) => {
        warn!(...);
        return Ok(ParseResult::New(...));
    }
}
```

Also, ditto re. warning.

**Comment by fitzgen** [comment]
@bors-servo r=emilio

Thanks! Review comments addressed.

**Pull request merged**